program/src/processor.rs

```diff
@@ -519,52 +521,82 @@ pub fn process_update_exchange_rate(
     lido.save(accounts.lido)
 }
 
+pub enum StakeType {
+    Stake,
+    Unstake,
+}
+
+impl std::fmt::Display for StakeType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StakeType::Stake => write!(f, "stake"),
+            StakeType::Unstake => write!(f, "unstake"),
+        }
+    }
+}
+
+pub struct WithdrawExcessOpts<'a, 'b> {
+    accounts: &'a WithdrawInactiveStakeInfo<'a, 'b>,
+    clock: &'a Clock,
+    rent: &'a Rent,
+    stake_history: &'a StakeHistory,
+    stake_account: &'a AccountInfo<'b>,
+}
+
 /// If a stake account contains more inactive SOL than needed to make it rent-exempt,
 /// withdraw the excess back to the reserve. Returns how much SOL was withdrawn.
-pub fn withdraw_excess_inactive_sol<'a, 'b>(
-    accounts: &WithdrawInactiveStakeInfo<'a, 'b>,
-    clock: &Clock,
-    rent: &Rent,
-    stake_history: &StakeHistory,
-    stake_account: &AccountInfo<'b>,
+/// When withdrawing from an unstake account, try to withdraw all the balance,
+/// leaving the stake account empty.
+pub fn withdraw_inactive_sol(
+    withdraw_excess_opts: &WithdrawExcessOpts,
     stake_account_seed: u64,
     stake_authority_bump_seed: u8,
+    stake_type: StakeType,
 ) -> Result<Lamports, ProgramError> {
-    let stake_account_rent = Lamports(rent.minimum_balance(stake_account.data_len()));
-    let stake = deserialize_stake_account(&stake_account.data.borrow())?;
+    let stake = deserialize_stake_account(&withdraw_excess_opts.stake_account.data.borrow())?;
     let stake_info = StakeAccount::from_delegated_account(
-        Lamports(stake_account.lamports()),
+        Lamports(withdraw_excess_opts.stake_account.lamports()),
         &stake,
-        clock,
-        stake_history,
+        withdraw_excess_opts.clock,
+        withdraw_excess_opts.stake_history,
         stake_account_seed,
     );
-    let excess_balance = Lamports(
-        stake_info
-            .balance
-            .inactive
-            .0
-            .saturating_sub(stake_account_rent.0),
-    );
+    let excess_balance = match stake_type {
+        StakeType::Stake => {
+            let stake_account_rent = Lamports(
+                withdraw_excess_opts
+                    .rent
+                    .minimum_balance(withdraw_excess_opts.stake_account.data_len()),
+            );
+            Lamports(
+                stake_info
+                    .balance
+                    .inactive
+                    .0
+                    .saturating_sub(stake_account_rent.0),
+            )
+        }
+        StakeType::Unstake => Lamports(stake_info.balance.inactive.0),
```

Comment:
If we go with the approach of always fully withdrawing a stake account, or withdrawing nothing at all, should we reflect that here, and only make the excess positive, when 100% of the balance is inactive?